### DISCUSSION TITLE
Change Logo and Branding

### DISCUSSION THREAD

**Discussion by Surgical71** [pglombardo/PasswordPusher#295] [Q&A] [answered]
Hi All

I just want to know if we are allowed to change the images in the container (Logo and Branding) and what would be the best way to have this done?

Idea is to give it a better look and feel not to pass the software off as my own.

**Comment by pglombardo** [comment]
> I just want to know if we are allowed to change the images in the container (Logo and Branding) and what would be the best way to have this done?

Definitely - I have no objections or requirements beyond what the open source licensing would enforce.  It's a Rails application so it would entail modifying ERB templates and changing the assets (images & CSS) in `./app/frontend`.

I've thought about making the branding configurable but it hasn't been implemented yet mostly because of scope.  Changing logo requires also attention to the CSS media queries which flip themes based on user preferences.

For example, you could replace the images in this header file to change the logo:
https://github.com/pglombardo/PasswordPusher/blob/master/app/views/shared/_header.html.erb#L6-L8

Then the application name and tag line are internationalized strings so you could substitute with just raw strings:
https://github.com/pglombardo/PasswordPusher/blob/master/app/views/shared/_header.html.erb#L12-L14

There is a feature request for this filed at #248

**Reply by speakdigital** [reply]
Thanks for this  - it's a really great piece of software. I also saw that you have more branding options in the pipeline which will be really handy. 

In the meantime I was able to login to the docker image and edit the erb files mentioned and put a logo in. I had to restart the docker container for it to show the change but I assume that's normal procedure?

However I'm struggling to see where to change the CSS. I updated app/frontend/css/bootstrap-<various>.css but to no avail. I can see by inspecting that the css file loaded is public/packs/css/application-9b0c6cdd.css but that appears to be a compiled file?

Thanks very much again

**Reply by pglombardo** [reply]
Hi @speakdigital - thanks for the kind words.  Those CSS files are compiled during the docker image compilation process with [bundle exec rails webpacker:compile](https://github.com/pglombardo/PasswordPusher/blob/master/containers/docker/pwpush-ephemeral/Dockerfile#L34).

After modifying those CSS files and testing with the raw source, I believe you would have to rebuild the container entirely to make sure all bases are covered.

Unfortunately I don't have a good solution for custom or modified CSS yet as Rails compiles/compacts the CSS and then fingerprints them.

A better way may be to late load a completely new CSS file in the application layouts (`app/views/layouts`) that would selectively override the originals although I haven't tested this yet.

Also note that [basic branding has been added](https://github.com/pglombardo/PasswordPusher/blob/master/Configuration.md#rebranding) recently.

Hopefully this helps - let me know how it goes.

**Reply by speakdigital** [reply]
@pglombardo thank you very much - this was really useful.

What I ended up doing was creating a speakdigital.css file in app/frontend/packs and then adding
`<%= stylesheet_pack_tag 'speakdigital', media: 'all', 'data-turbolinks-track': 'reload' %>`
to app/views/layouts/application.html.erb

So we now have [https://password.speakdigital.co.uk](https://password.speakdigital.co.uk) all purpled up and slightly tweaked.

Thanks again for this great tool.

**Reply by pglombardo** [reply]
Nice! 😎

**Comment by pglombardo** [comment]
For others coming to this discussion, branding is now supported.  Documentation is here: https://github.com/pglombardo/PasswordPusher/blob/master/Configuration.md#rebranding

**Comment by pglombardo** [comment]
Hi all - beyond the branding support that was added last year, I put together some [instructions on how to download and apply any theme](https://github.com/pglombardo/PasswordPusher/issues/762#is) to Password Pusher.  It's a short term work around but works well.

![Screenshot 2023-02-04 at 11 07 40](https://user-images.githubusercontent.com/395132/216762828-90ad3ac6-2f11-402c-9a90-8ba2c00e04ba.png)

I hope this helps!

**Comment by pglombardo** [comment, marked as answer]
Hi all - I'm updating the answer here with the latest information.

As of v1.26.10 Password Pusher now

* comes with [26 themes](https://github.com/pglombardo/PasswordPusher/blob/master/Themes.md#theme-gallery) bundled in selectable by environment variable
* also supports [custom CSS](https://github.com/pglombardo/PasswordPusher/blob/master/Configuration.md#how-to-add-custom-css)
* supports [entirely custom themes](https://github.com/pglombardo/PasswordPusher/blob/master/Configuration.md#adding-an-entirely-new-theme-from-scratch)
* has the ability to be [re-branded](https://github.com/pglombardo/PasswordPusher/blob/master/Configuration.md#rebranding) with logos, tag lines and more

With this framework and foundation, we'll be able to make sure that future changes don't break any existing user customisations.

If there are any questions or issues, please let me know!
